File: jva/src/clientAis/MainClientAi.java
package clientAis;


import clientAis.cli.CommandLineParser;
import clientAis.communications.MessageSerializer;
import clientAis.communications.ServerCommander;
import clientAis.communications.ServerReceiver;
import clientAis.communications.game_data.GameState;
import clientAis.dynamic_data.DataPacket;
import clientAis.games.GameManager;
import clientAis.implementations.Bot;
import clientAis.implementations.basic_minion_wielder.BasicMinionWielder;
import clientAis.implementations.basic_single_mind.BasicSingleMind;
import clientAis.implementations.challenge_defense.ChallengeDefense;
import clientAis.implementations.closest_spawner.ClosestSpawner;
import clientAis.implementations.dummy.Dummy;
import clientAis.implementations.go_middle.GoMiddle;
import clientAis.implementations.mindless_chase.MindlessChase;
import clientAis.implementations.multidefense.MultiDefense;
import clientAis.implementations.relentless_attacker.RelentlessAttacker;
import clientAis.implementations.threat_level_defender.ThreatLevelDefender;
import clientAis.networking.ClientConnector;
import utils.timer.LambdaTimerTaskHelper;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.Timer;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;

public class MainClientAi {

    public static final String DEFAULT_SERVER_IP = "127.0.0.1";
    public static final String DEFAULT_SERVER_PORT = "52124";
    public static final String DEFAULT_AI_NAME = MultiDefense.class.getName();
    public static final double SECOND_BETWEEN_AI_FRAME = 0.1;
    public static final long REFRESH_PERIOD_MS = (int)(SECOND_BETWEEN_AI_FRAME*1000);

    private static final Map<String, Bot> BOT_IMPLEMENTATIONS = new HashMap<>();
    static {
        addBotImplementation(new Dummy());
        addBotImplementation(new GoMiddle());
        addBotImplementation(new ClosestSpawner());
        addBotImplementation(new MindlessChase());
        addBotImplementation(new ChallengeDefense());
        addBotImplementation(new BasicSingleMind());
        addBotImplementation(new BasicMinionWielder());
        addBotImplementation(new ThreatLevelDefender());

        addBotImplementation(new RelentlessAttacker());
        addBotImplementation(new MultiDefense());
    }

    private static Optional<DataPacket> previousInputOpt = Optional.empty();

    private static void addBotImplementation(Bot bot) {
        BOT_IMPLEMENTATIONS.put(bot.getClass().getName(), bot);
    }

    public static void main(String[] args) {
        final CommandLineParser commandLineParser = new CommandLineParser(args);

        final String serverIp = commandLineParser.findArgument("serverIp")
                .orElse(DEFAULT_SERVER_IP);
        final String serverPortString = commandLineParser.findArgument("serverPort")
                .orElse(DEFAULT_SERVER_PORT);
        final int serverPort = Integer.parseInt(serverPortString);
        final String aiName = commandLineParser.findArgument("aiName")
                .orElse(DEFAULT_AI_NAME);

        ClientConnector.connect(serverIp, serverPort, communicationHandler -> {
            System.out.println("Connected to server at " + serverIp + ":" + serverPort);

            final MessageSerializer messageSerializer = new MessageSerializer();
            final ServerCommander serverCommander = new ServerCommander(communicationHandler, messageSerializer);

            final GameManager gameManager = new GameManager();

            final ServerReceiver serverReceiver = new ServerReceiver(communicationHandler, gameManager, messageSerializer);
            serverReceiver.startAsync();

            System.out.println("Started games evaluation");

            final Bot bot = BOT_IMPLEMENTATIONS.get(aiName);

            AtomicInteger lastAiGameStateVersion = new AtomicInteger(-1);

            Timer timer = new Timer();
            timer.scheduleAtFixedRate(new LambdaTimerTaskHelper(() -> {
                if(communicationHandler.isClosed) {
                    timer.cancel();
                }

                gameManager.gameState.ifPresent(gameState -> gameManager.currentPlayerId.ifPresent(currentPlayerId -> {
                    if(gameState.frameCount > lastAiGameStateVersion.get()) {
                        lastAiGameStateVersion.set(gameManager.gameState.get().frameCount);
                    }
                    final double timeBeforeRunningBot = System.nanoTime()/1000000.0;
                    runBot(bot, currentPlayerId, gameState, serverCommander);
                    final double timeAfterRunningBot = System.nanoTime()/1000000.0;
                    final double deltaTime = timeAfterRunningBot - timeBeforeRunningBot;
                    System.out.println("Frame " + gameState.frameCount + " took " + String.format("%,.3f", deltaTime) + "ms. CPU " + (int)(100*deltaTime/REFRESH_PERIOD_MS) + "%.");

                }));
            }), 0, REFRESH_PERIOD_MS);
        });
    }

    private static void runBot(
            final Bot bot,
            final String currentPlayerId,
            final GameState gameState,
            final ServerCommander serverCommander) {
        final DataPacket input = new DataPacket(gameState, currentPlayerId, previousInputOpt);
        final Consumer<ServerCommander> consumer = bot.exec(input);
        previousInputOpt = Optional.of(input);
        consumer.accept(serverCommander);
    }
}

File: py/clientAis/mainClientAi.py
import sys
import time
from typing import Callable, Dict

from ais.Artificial import Artificial
from ais.SingleMindClosest.SingleMindClosest import SingleMindClosest
from ais.Voider.Voider import Voider
from clientAis.ais.PropagatingVision.PropagatingVision import PropagatingVision
from networking.ClientConnector import ClientConnector
from communications.MessageSerializer import MessageSerializer
from games.GameManager import GameManager
from networking.CommunicationHandler import CommunicationHandler
from communications.ServerCommander import ServerCommander
from communications.ServerReceiver import ServerReceiver
from utils import arrays

commandLineOptions = {key: value for key, value in zip(arrays.soft_accessor(sys.argv, 1, None, 2), arrays.soft_accessor(sys.argv, 2, None, 3))}

DEFAULT_SERVER_IP = "127.0.0.1"
DEFAULT_SERVER_PORT = 52124
DEFAULT_AI_NAME = "singleMindClosest"
SECOND_BETWEEN_AI_FRAME = 0.5

serverIp = commandLineOptions.get("serverIp") or DEFAULT_SERVER_IP
serverPort = int(commandLineOptions.get("serverPort") or DEFAULT_SERVER_PORT)
aiName = commandLineOptions.get("aiName") or DEFAULT_AI_NAME

ais: Dict[str, Callable[[ServerCommander], Artificial]] = {
    "voider": lambda serverCommander: Voider(serverCommander),
    "singleMindClosest": lambda serverCommander: SingleMindClosest(serverCommander),
    "propagatingVision": lambda serverCommander: PropagatingVision(serverCommander)
}

def clientConnectorCallBack(communicationHandler: CommunicationHandler):
    print(f"Connected to server at {serverIp}:{serverPort}")

    gameManager = GameManager()

    messageSerializer = MessageSerializer()
    serverCommander = ServerCommander(communicationHandler, messageSerializer)

    artificial = ais[aiName](serverCommander)

    serverReceiver = ServerReceiver(communicationHandler, gameManager, messageSerializer)

    serverReceiver.startAsync()

    print("Starting games evaluation")

    lastAiGameStateVersion = -1

    try:
        while not communicationHandler.closed:
            timeBeforeFrame = time.time()

            if gameManager.gameState is not None and gameManager.gameState.frameCount > lastAiGameStateVersion:
                lastAiGameStateVersion = gameManager.gameState.frameCount
                artificial.frame(gameManager.gameState, gameManager.currentPlayerId)

            timeAfterFrame = time.time()

            frameTimeInSecond = timeAfterFrame - timeBeforeFrame

            if frameTimeInSecond < SECOND_BETWEEN_AI_FRAME:
                print(f"Frame {0 if gameManager.gameState is None else gameManager.gameState.frameCount} took {frameTimeInSecond:.3f}s. Sleeping {SECOND_BETWEEN_AI_FRAME - frameTimeInSecond:.3f}s.")
                time.sleep(SECOND_BETWEEN_AI_FRAME - frameTimeInSecond)
            else:
                print(f"Frame {0 if gameManager.gameState is None else gameManager.gameState.frameCount} took {frameTimeInSecond:.3f}s. Not sleeping")
    except:
        raise
    finally:
        serverReceiver.stop()

ClientConnector(serverIp, serverPort, clientConnectorCallBack)
